juju/model/__init__.py

```diff
@@ -3205,14 +3246,87 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
-                raise asyncio.TimeoutError("Timed out waiting for model:\n" + busy)
+            if timeout_ is not None and datetime.now() - start_time > timeout_:
+                raise asyncio.TimeoutError(
+                    "\n  ".join(["Timed out waiting for model:", *busy])
+                )
             if last_log_time is None or datetime.now() - last_log_time > log_interval:
-                log.info("Waiting for model:\n  " + busy)
+                log.info("\n  ".join(["Waiting for model:", *busy]))
                 last_log_time = datetime.now()
             await asyncio.sleep(check_freq)
 
+    async def new_wait_for_idle(
+        self,
+        apps: Iterable[str] | None = None,
+        raise_on_error: bool = True,
+        raise_on_blocked: bool = False,
+        wait_for_active: bool = False,
+        timeout: float | None = 10 * 60,
+        idle_period: float = 15,
+        check_freq: float = 0.5,
+        status: str | None = None,
+        wait_for_at_least_units: int | None = None,
+        wait_for_exact_units: int | None = None,
+    ) -> None:
+        """Wait for applications in the model to settle into an idle state.
+
+        arguments match those of .wait_for_idle exactly.
+        """
+        if not isinstance(wait_for_exact_units, (int, type(None))):
+            raise ValueError(f"Must be an int or None, got {wait_for_exact_units=}")
+
+        if isinstance(wait_for_exact_units, int) and wait_for_exact_units < 0:
+            raise ValueError(f"Must be >=0, got {wait_for_exact_units=}")
+
+        if wait_for_active:
+            warnings.warn(
+                "wait_for_active is deprecated; use status",
+                DeprecationWarning,
+                stacklevel=3,
+            )
+            status = "active"
+
+        wait_for_units = (
+            wait_for_at_least_units if wait_for_at_least_units is not None else 1
+        )
+
+        if isinstance(apps, (str, bytes, bytearray, memoryview)):
+            raise TypeError(f"apps must be a Iterable[str], got {apps=}")
+
+        apps = frozenset(apps or self.applications)
+
+        if any(not isinstance(o, str) for o in apps):
+            raise TypeError(f"apps must be a Iterable[str], got {apps=}")
+
+        started = time.monotonic()
+        deadline = None if timeout is None else started + timeout
+
+        async def status_on_demand():
+            while True:
+                yield _idle.check(
+                    await self.get_status(),
+                    apps=apps,
+                    raise_on_error=raise_on_error,
+                    raise_on_blocked=raise_on_blocked,
+                    status=status,
+                )
+
+        async for done in _idle.loop(
+            status_on_demand(),
+            apps=apps,
+            wait_for_exact_units=wait_for_exact_units,
+            wait_for_units=wait_for_units,
+            idle_period=idle_period,
+        ):
```

Comment:
Suggestion: simplify this for noobs like me ... btw at first I thought this could be simplified without changing anything else, but after making it to `_idle.loop` I realised that my original simplification was totally incorrect, which suggests that the current version is a bit tricky to follow ... the below suggestion assumes a later suggestion to refactor `_idle.loop` to use a class, which I think makes things a lot easier to follow -- it certainly does for me
```suggestion
        timing_checker = _idle.TimingChecker()
        while True:
            full_status = await self.get_status()
            check_status = _idle.check(
                raw_status,
                apps=apps,
                raise_on_error=raise_on_error,
                raise_on_blocked=raise_on_blocked,
                status=status,
            )
            done = timing_checker.check(
                check_status,
                apps=apps,
                wait_for_exact_units=wait_for_exact_units,
                wait_for_units=wait_for_units,
                idle_period=idle_period,
            )
```